### DISCUSSION TITLE
[More question than issue] How to set platform-dependend include paths for JNI library?

### DISCUSSION THREAD

**Discussion by blacklion** [nokeedev/gradle-native#535] [Q&A] [answered]
Sorry, I can not find any forum, mailing list or other community support for your project, so I ask here.

I need to link JNI library to system-depended libraries and include path (!) for JNI library. 

This works:
```
library.variants.configureEach {
  sharedLibrary {
    linkTask.configure {
      if (targetMachine.operatingSystemFamily.windows) {
        linkerArgs.add(rootProject.file('libs/win32-x86-64/depend-imp.lib').absolutePath)
      } else if (targetMachine.operatingSystemFamily.linux) {
        linkerArgs.add('-ldepend')
      } else if (targetMachine.operatingSystemFamily.macOS) {
        linkerArgs.add('-ldepend')
      } else {
        logger.warn('Unknown operating system: ' + targetMachine.operatingSystemFamily.name + ', hope it is GCC-compatible')
        linkerArgs.add('-L/usr/local/lib')
        linkerArgs.add('-ldepend')
      }
    }
  }
}
```

This doesn't work and I can not find any documentation how to do this:

```
library.variants.configureEach {
  sharedLibrary {
    compileTask.configure {
      if (targetMachine.operatingSystemFamily.windows) {
        headerSearchPaths.add(rootProject.file('libs/header').absolutePath)
      } else if (targetMachine.operatingSystemFamily.linux) {
      } else if (targetMachine.operatingSystemFamily.macOS) {
      } else {
        logger.warn('Unknown operating system: ' + targetMachine.operatingSystemFamily.name + ', hope it is GCC-compatible')
        headerSearchPaths.add('/usr/local/include')
      }
    }
  }
}
```

It complains:

```
* What went wrong:
Could not determine the dependencies of task ':project-name:assemble'.
> Could not create domain object 'windowsX86-64' (JniLibraryInternal)
```

**Comment by lacasseio** [comment, marked as answer]
Sorry for missing your issue/question, there's a [discussion tab](https://github.com/nokeedev/gradle-native/discussions) on the project (I will move the issue there). You can also use the Gradle community Slack, there's a `#native` channel.

It's a great question and the answer is to use `includes.from('<path>')`. There's quite a legacy behind why the documentation only shows `headerSearchPaths`. The `includes` property is a public-ish `ConfigurableFileCollection` property that comes from Gradle core types which we currently use. The unfortunate downside of the single `includes` property is the lack of information to distinguish between a header search path that is an include root (directory) vs a framework on macOS. Ideally, both are header search paths. I will expose more of the Gradle core properties for the time being until we can finish the implementation of our types. The Gradle core types will go through a deprecation cycle as lots of users are dependent upon them.

Your thinking was right. You can always use `compilerArgs.addAll('-I', <path>)`. However, it's usually best to avoid straight compiler arguments for things that are modelled as we may add some optimization or better reporting around them. Despite that, we will most likely notify the user when flags are used instead of the model later on.

**Reply by btbouwens** [reply]
OK, let me add a newbee question to this. I know Java, I know JNI, but gradle is still kind of new to me.

So my question is: where do I put the  `includes.from('<path>')`? I think in a compile task in the build.gradle.kts, but then I'm lost..

Actually the documentation mentions many more settings, it would be nice to see more about how to use them.

Cheers - Bram
